File: C++/Graphs/mother-vertex.cpp


// Given a graph, find the mother vertex

// Mother vertex -  a vertex from where you can reach all the other vertices

// Quick tip : The Mother vertex is the last vertex from a DFS

#include <iostream>
#include <list>
#include <stack>
using namespace std;

int V;
list<int> *adj;

void addEdge(int v, int w) {
    adj[v].push_back(w);
}

void DFS(int start, bool *visited) {

    // Top value from the stack
    int topValue; 

    // We will use a stack instead of recursion
    stack<int> st;

    // Add it to the stack so we can start
    st.push(start);

    while(!st.empty()) {

        topValue = st.top();
        st.pop();
        // Mark the current node as visited
        visited[topValue] = true;

        // Look for adjacent nodes
        for(auto it : adj[topValue]) {
            if(!visited[it]) {
                st.push(it);
            }
        }

    }

}

int findMotherVertex() {

    // We will do a DFS for each node
    // And will store in a variable the last vertex finished in DFS

    // Keep track of each vertex
    bool *visited = new bool[V];
    for(int i = 0; i < V;i++)
        visited[i] = false;

    // Here we store the last vertex finished in DFS
    int v;

    for(int i =0; i < V;i++) {
        if(!visited[i]) {
            DFS(i,visited);
            // Store the last index
            v = i;
        }
    }

    // Reset the array
    for(int i = 0; i < V;i++)
    visited[i] = false;
    
    // Start a DFS from the node "v"
    DFS(v,visited);

    // Check if all nodes are visited
    for(int i=0; i < V;i++) 
        if(!visited[i])
            return -1;

    return v;
}

int main () {

    V = 7;
    adj = new list<int>[V];
    addEdge(0, 1);
    addEdge(0, 2);
    addEdge(1, 3);
    addEdge(4, 1);
    addEdge(6, 4);
    addEdge(5, 6);
    addEdge(5, 2);
    addEdge(6, 0);

    cout << findMotherVertex() << "\n"; // 5
    return 0;

}

File: Python/Graphs/mother-vertex.py

#* Find mother vertex in a graph

# Mother vertex - is a vertex from where you can reach every other vertex in the graph

# The mother vertex is the last index found after a DFS 

#defaultdict - returns an dictionary-like object
from collections import defaultdict

class Graph:
    def __init__(self,V):
        self.V = V
        self.adj = defaultdict(list) # Basically create a list of lists
    
    # Add edge between 2 nodes
    def addEdge(self,v,w):
        self.adj[v].append(w)

    def DFS(self,start,visited):
        # Mark the current node visited
        visited[start]  = True

        # Search for adjacent nodes
        for i in self.adj[start]: # Search in that particular list
            if(not visited[i]):
                self.DFS(i,visited)

    def findMotherVertex(self):
        
        # Keep track of every node in the graph
        visited = [False for i in range(self.V)]

        # Find the last node after DFS
        for i in range(self.V):
            # If we haven't visited the node yet
            if(not visited[i]):
                # Start a DFS 
                self.DFS(i,visited)
                # Assume that is the mother vertex (the last vertex found after a DFS)
                v = i

        # Reset the visited list
        visited = [False] * self.V

        # Check if we can reach every node by starting a DFS from this "v" node
        self.DFS(v,visited) # If this is the mother vertex, then every other node should be visited
        for i in range(self.V):
            if(not visited[i]): # If a node has not been visited, it means there is no mother vertex
                return -1
        # If we got to this point, it means a mother vertex exists
        return v
        

# Create a graph
g = Graph(7)
g.addEdge(0, 1)
g.addEdge(0, 2)
g.addEdge(1, 3)
g.addEdge(4, 1)
g.addEdge(6, 4)
g.addEdge(5, 6)
g.addEdge(5, 2)
g.addEdge(6, 0)

print g.findMotherVertex() # 5